src/grammar_optimiser/analyze_compressions.jl

```diff
@@ -0,0 +1,153 @@
+"""
+    generate_stats(d, compressed_AST)
+
+Compression Analysis. Analyzes 1 AST to see how many times each compression was used.
+# Arguments
+- `d::Dict`: the global dictionary (key: node_id, value: namedTuple(compressiond_id, parent_id, child_nr, type, [children]))
+- `compressed_AST::Vector{String}`: a list of assign-statements ["assign(A, X)", assign(B, Y), ...]
+# Result 
+- `c_info::Dict{Int64, NamedTuple{(:size, :occurences), <:Tuple{Int64,Int64}}}`: an dict(key: compression_id, value: Tuple(size, # occurences))) 
+"""
+function generate_stats(d, compressed_AST)
+    # (key: subtree ID, value: NamedTuple([subtree node IDs], # occurences))
+    # c_info = Dict{Int64, NamedTuple{:size, :occurences}{Int64, Int64}}()
+    c_info = Dict{Int64, NamedTuple{(:size, :occurences), <:Tuple{Int64,Int64}}}()
+
+
+    for assign in compressed_AST
+
+        # parse the compression node id
+        node_id = nothing
+
+        m = match(r"\((\d+),", assign)
+
+        @assert m !== nothing
+        @assert length(m.captures) == 1
+        node_id = parse(Int64, m.captures[1])
+
+        # find all the compressions of that node
+        C = d[node_id].comp_id
+
+        # increment the counter if the compression C has been used already
+        if haskey(c_info, C)
+            c_info[C] = (size = c_info[C].size, occurences = c_info[C].occurences + 1)
+
+        # initialize the counter for the first usage of the compression
+        else
+            c_info[C] = (size = getCompressionSize(d, C), occurences = 1)
+        end
+    end
+
+    for (C, v) in c_info
+        # the sum of occurences of all nodes of a compression must be exactly divisible by the compression's size
+        @assert (mod(v.occurences, v.size) == 0) || (v.size == 0)
+        c_info[C] = (size = v.size, occurences = trunc(Int, v.occurences / v.size))
+    end
+
+    return c_info
+end
+
+"""
+    compare(rn₁, rn₂)
+
+Compares two RuleNodes. Returns true if they are equal, false otherwise.
+# Arguments
+- `rn₁::RuleNode`: a RuleNode
+- `rn₂::RuleNode`: a RuleNode
+# Result
+- `Bool`: true if the RuleNodes are equal, false otherwise
+"""
+function compare(rn₁, rn₂)::Bool  
```

Comment:
This already exists in `HerbCore`: https://github.com/Herb-AI/HerbCore.jl/blob/2d5dcc60adf2d7067cefa5a34ac36895c1715acc/src/rulenode.jl#L112